src/System.Net.Utilities/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -2,33 +2,291 @@
+                    cts.Cancel();
+                    if (finished != receiveTask)
+                    {
+                        sw.Stop();
+                        throw new PingException(SR.net_ping_reply_timeout);
+                    }
+
+                    SocketReceiveFromResult receiveResult = receiveTask.GetAwaiter().GetResult();
+                    int bytesReceived = receiveResult.ReceivedBytes;
+                    if (bytesReceived - ipHeaderLength < IcmpHeaderLengthInBytes)
+                    {
+                        continue; // Not enough bytes to reconstruct IP header + ICMP header.
+                    }
+
+                    byte type, code;
+                    unsafe
+                    {
+                        fixed (byte* bytesPtr = receiveBuffer)
+                        {
+                            int icmpHeaderOffset = ipHeaderLength;
+                            IcmpHeader receivedHeader = *((IcmpHeader*)(bytesPtr + icmpHeaderOffset)); // Skip IP header.
+                            type = receivedHeader.Type;
+                            code = receivedHeader.Code;
+
+                            if (identifier != receivedHeader.Identifier
+                                || type == Icmpv4MessageConstants.EchoRequest
+                                || type == Icmpv6MessageConstants.EchoRequest) // Echo Request, ignore
+                            {
+                                continue;
+                            }
+                        }
+                    }
+
+                    sw.Stop();
+                    long roundTripTime = sw.ElapsedMilliseconds;
+                    int dataOffset = ipHeaderLength + IcmpHeaderLengthInBytes;
+                    // We want to return a buffer with the actual data we sent out, not including the header data.
+                    byte[] dataBuffer = new byte[bytesReceived - dataOffset];
+                    Array.Copy(receiveBuffer, dataOffset, dataBuffer, 0, dataBuffer.Length);
+
+                    IPStatus status = isIpv4
+                                        ? Icmpv4MessageConstants.MapV4TypeToIPStatus(type, code)
+                                        : Icmpv6MessageConstants.MapV6TypeToIPStatus(type, code);
+
+                    return new PingReply(address, options, status, roundTripTime, dataBuffer);
+                }
+
+                // We have exceeded our timeout duration, and no reply has been received.
+                sw.Stop();
+                throw new PingException(SR.net_ping_reply_timeout);
+            }
+        }
+
+        private async Task<PingReply> SendWithPingUtility(IPAddress address, byte[] buffer, int timeout, PingOptions options)
+        {
+            bool isIpv4 = address.AddressFamily == AddressFamily.InterNetwork;
+            string pingExecutable = isIpv4 ? s_ipv4PingFile : s_ipv6PingFile;
+
+            StringBuilder sb = new StringBuilder();
```

Comment:
Yeah, agreed that this could use some additional testing (other than the success/failure of the whole command). When I restructure the tests separately, I'll add a test to verify the behavior of these. We need to validate that

* Only one ping is sent per invocation
* The size of the packet is actually respected

We can do both with additional parsing in the test.